media-libs/mesa: add rusticl use flag

darkbasic opened the pull request:
Rusticl is a much better alternative to clover, which will be gone soon: https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/19385

This PR tries to add a rusticl use flag to mesa, but there are several blockers:
- ~~`dev-libs/libclc` needs a `spirv` use flag (also needed for Ray Tracing)~~ _which I added_
- ~~`dev-util/rust-bindgen` requires `FEATURES="-network-sandbox"` in order to build until we will start to package crates individually~~ ~~_turned out existing cargo.eclass was more than capable enough_~~ already merged by someone else
https://bugs.gentoo.org/680984
https://wiki.gentoo.org/wiki/Project:Rust
https://github.com/kentnl-gentoo-rust/rust-dev-overlay/blob/master/dev-rust/bindgen/bindgen-0.53.1.ebuild
- ~~`-Dopencl-spirv` is unable to find LLVM unless you symlink `/usr/lib/clang` to `/usr/lib/llvm/15/lib64/clang`.
mesa developers think it's a packaging issue due to slotting, but they're currently looking for a solution anyway:
https://bugs.gentoo.org/880615~~ fixed: https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/19617
- ~~`PKG_CONFIG_LIBDIR` needs to point to `/usr/lib/llvm/15/lib64/pkgconfig` or `LLVMSPIRVLib` won't be found.~~ _I've hardcoded it for the `rusticl` use flag, so that **MIGHT** be enough._

I've also added use flags for all the possible OpenCL backends in mesa [clover, rusticl, lpcl (sw)] but we should decide which ones we want to keep.

darkbasic commented:
I switch git repo and branch to allow broader testing with radeonsi until it's ready to be merged: https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/19232

It works very well, even on ppc64le:
https://pastebin.com/qp3nngpA
https://openbenchmarking.org/result/2211096-DARK-DARKTAB52&cts=1

thesamesam commented:
How is spirv a blocker if you've done it here?

I suspect you can just use cargo-ebuild like every other Rust build for bindgen. If not, and this shouldn't be needed, we can create a vendor tarball.

darkbasic commented:
> How is spirv a blocker if you've done it here?

I misphrased my sentence, the others are real blocker in this case I just meant that we also need the other commit to land.

> I suspect you can just use cargo-ebuild like every other Rust build for bindgen. If not, and this shouldn't be needed, we can create a vendor tarball.

There aren't many rust packages in Gentoo and I'm not aware of what are the current guidelines, so I will need some help or at least some inputs for this.

darkbasic commented:
I've looked at the cargo eclass and it turned out it was more than capable enough to create an ebuild with the need for `FEATURES="-network-sandbox"`.

thesamesam commented on media-libs/mesa/mesa-9999.ebuild:
```diff
@@ -188,6 +194,12 @@ BDEPEND="
 		>=sys-devel/gcc-4.6
 	)
+	rusticl? (
+		>=dev-lang/rust-1.59.0
```
You always want virtual/rust, not dev-lang/rust, to allow people to use rust-bin.

FireBurn commented:
The libclc spirv stuff is now in ::gentoo thanks to @mgorny so that can be dropped now

darkbasic commented on media-libs/mesa/mesa-9999.ebuild:
```diff
@@ -13,7 +13,8 @@ DESCRIPTION="OpenGL-like graphic library for Linux"
 
 if [[ ${PV} == 9999 ]]; then
-	EGIT_REPO_URI="https://gitlab.freedesktop.org/mesa/mesa.git"
+	EGIT_REPO_URI="https://gitlab.freedesktop.org/karolherbst/mesa.git"
```
We don't: it's here just for convenience to allow developers with radeon cards to test it, but there is already upstream support for Intel. I can remove it now or once we're ready to merge this.

thesamesam replied:
That's fine, just put a TODO/FIXME above it.

darkbasic commented on dev-util/rust-bindgen/rust-bindgen-0.61.0.ebuild:
```diff
@@ -0,0 +1,121 @@
+		sys-devel/clang:13
+	)
+"
+RDEPEND="${DEPEND}"
```
Both Arch Linux (https://github.com/archlinux/svntogit-community/blob/packages/rust-bindgen/trunk/PKGBUILD) and Debian (https://packages.debian.org/sid/bindgen) have depends (which should be runtime AFAIK) on clang.

darkbasic commented on dev-util/rust-bindgen/rust-bindgen-0.61.0.ebuild:
```diff
@@ -0,0 +1,121 @@
+DESCRIPTION="A binding generator for the rust language"
+HOMEPAGE="https://github.com/servo/rust-bindgen"
+
+LICENSE="BSD"
```
This is the output:
```
(MIT OR Apache-2.0) AND Unicode-DFS-2016 (1): unicode-ident
Apache-2.0 (2): clang-sys, thread-id
Apache-2.0 OR Apache-2.0 WITH LLVM-exception OR MIT (1): wasi
Apache-2.0 OR MIT (53): autocfg, bitflags, cc, cexpr, cfg-if, clap, clap_lex, diff, either, env_logger, env_logger, getrandom, glob, hashbrown, hermit-abi, humantime, indexmap, lazy_static, lazycell, libc, log, log, minimal-lexical, os_str_bytes, peeking_take_while, ppv-lite86, proc-macro2, quote, rand, rand, rand, rand_chacha, rand_core, rand_core, rand_core, rand_hc, regex, regex, regex-syntax, regex-syntax, remove_dir_all, rustc-hash, shlex, syn, tempdir, tempfile, thread_local, unicode-width, vec_map, version_check, winapi, winapi-i686-pc-windows-gnu, winapi-x86_64-pc-windows-gnu
BSD-3-Clause (3): bindgen, bindgen, bindgen-cli
Custom License File (1): fuchsia-cprng
ISC (3): libloading, libloading, rdrand
MIT (16): ansi_term, atty, block, clap, kernel32-sys, malloc_buf, nom, objc, redox_syscall, strsim, strsim, textwrap, textwrap, which, winapi, winapi-build
MIT OR Unlicense (8): aho-corasick, aho-corasick, memchr, memchr, quickcheck, termcolor, utf8-ranges, winapi-util
N/A (4): bindgen-integration, bindgen-tests, quickchecking, tests_expectations
```

I've set it to `LICENSE="Apache-2.0 Apache-2.0-with-LLVM-exceptions BSD-3-Clause ISC MIT Unlicense"`

mgorny commented:
Quick question: right now spirv-llvm-translator is keyworded on amd64 only, and apparently fails a few tests on x86 (according to dirty testing by @thesamesam). Is this intended to work on x86 as well, or should I just mask `libclc[spirv]` on x86 for the time being?

darkbasic commented:
> Hey, you didn't answer my question about "lpcl" and adding rusticl twice to the IUSE

Did you forget to send your review? GitHub has this bad habit of requiring one additional confirmation and it's to oversee it.
I have no idea what you asked about the `lpcl` use, but the duplicate `rusticl` use is an oversight.

darkbasic commented:
> It's also likely that clover will be removed from mesa once the radeonsi stuff is merged for rusticl

It won't be so easy: R600 and function calls need to land as well and it could take months. I'm fine removing clover support before mesa does, but it's not up to me to decide.

darkbasic commented:
> Is this intended to work on x86 as well, or should I just mask libclc[spirv] on x86 for the time being?

You'd better ask Karol Herbst on this. I've tested it on ppc64 so far and obviously it works on amd64 as well. I've added the `~x86` keyword on `dev-util/rust-bindgen` because Debian does but I'm unsure about libclc.

FireBurn commented:
Let me have a go at making spriv-llvm-translator mulitlib, I'll see about trying to fix the pkgconfig too. In my overlay I've only gotten 64bit builds of mesa working with rusticl, possibly because of the missing translator, or possibly because rust doesn't like the way portage is calling it for 32bit binaries. Will post the output here either way

FireBurn commented:
I've adjusted the spirv-llvm-translator to use just the usual prefix rather than installing into the llvm area and make it multilib ( https://github.com/FireBurn/Overlay/blob/master/dev-util/spirv-llvm-translator/spirv-llvm-translator-15.0.0.ebuild )

FireBurn commented:
Here's what my mesa build says when trying to run rust as 32bit:

 * abi_x86_32.x86: running multilib-minimal_abi_src_configure
meson setup --libdir lib --localstatedir /var/lib --prefix /usr --sysconfdir /etc --wrap-mode nodownload --build.pkg-config-path /usr/share/pkgconfig:/usr/share/pkgconfig --pkg-config-path /usr/share/pkgconfig:/usr/share/pkgconfig --native-file /var/tmp/portage/media-libs/mesa-9999/temp/meson.i686-pc-linux-gnu.x86.ini -Db_pch=false -Dwerror=false --buildtype plain -Dplatforms=x11,wayland -Dgallium-nine=true -Dgallium-va=enabled -Dva-libs-path=/usr/lib/va/drivers -Dgallium-vdpau=disabled -Dgallium-xa=disabled -Dgallium-opencl=disabled -Dgallium-rusticl=true -Drust_std=2021 -Dvulkan-layers=device-select,overlay -Dbuild-tests=false -Dglx=dri -Dshared-glapi=enabled -Ddri3=enabled -Degl=enabled -Dgbm=enabled -Dglvnd=true -Dgles1=enabled -Dgles2=enabled -Dllvm=enabled -Dlmsensors=enabled -Dosmesa=false -Dselinux=false -Dlibunwind=disabled -Dzstd=enabled -Dsse2=true -Dvalgrind=disabled -Dvideo-codecs=h264dec,h264enc,h265dec,h265enc,vc1dec -Dgallium-drivers=radeonsi,swrast,zink -Dvulkan-drivers=amd --buildtype plain -Db_ndebug=true /var/tmp/portage/media-libs/mesa-9999/work/mesa-9999 /var/tmp/portage/media-libs/mesa-9999/work/mesa-9999-abi_x86_32.x86
The Meson build system
Version: 0.64.0
Source dir: /var/tmp/portage/media-libs/mesa-9999/work/mesa-9999
Build dir: /var/tmp/portage/media-libs/mesa-9999/work/mesa-9999-abi_x86_32.x86
Build type: native build
Program python3 found: YES (/var/tmp/portage/media-libs/mesa-9999/temp/python3.10/bin/python3)
Project name: mesa
Project version: 23.0.0-devel
C compiler for the host machine: clang-15 -m32 -mfpmath=sse (clang 15.0.4 "clang version 15.0.4")
C linker for the host machine: clang-15 -m32 -mfpmath=sse ld.lld 15.0.4
C++ compiler for the host machine: clang++-15 -m32 -mfpmath=sse (clang 15.0.4 "clang version 15.0.4")
C++ linker for the host machine: clang++-15 -m32 -mfpmath=sse ld.lld 15.0.4
Host machine cpu family: x86
Host machine cpu: i686
Checking if "-mtls-dialect=gnu2" runs: DID NOT COMPILE
Program glslangValidator found: YES (/usr/bin/glslangValidator)
Found pkg-config: /usr/bin/x86_64-pc-linux-gnu-pkg-config (1.8.0)
Found CMake: /usr/bin/cmake (3.24.3)
Run-time dependency libomxil-bellagio found: NO (tried pkgconfig and cmake)
Run-time dependency libtizonia found: NO (tried pkgconfig and cmake)
Run-time dependency libtizplatform found: NO (tried pkgconfig and cmake)
Run-time dependency tizilheaders found: NO (tried pkgconfig and cmake)
Run-time dependency libva found: YES 1.16.0
Header "va/va.h" has symbol "VASurfaceAttribDRMFormatModifiers" with dependency libva: YES 
Compiler for language rust for the build machine not found.

../mesa-9999/meson.build:935:2: ERROR: Rust compiler rustc -C linker=clang-15 -C link-arg=-m32 -C link-arg=-mfpmath=sse can not compile programs.

error: linking with `clang-15` failed: exit status: 1
  |
  = note: "clang-15" "-m64" "/var/tmp/portage/media-libs/mesa-9999/temp/rustc5oecoZ/symbols.o" "/var/tmp/portage/media-libs/mesa-9999/work/mesa-9999-abi_x86_32.x86/meson-private/rusttest.sanity.bf099191-cgu.0.rcgu.o" "/var/tmp/portage/media-libs/mesa-9999/work/mesa-9999-abi_x86_32.x86/meson-private/rusttest.sanity.bf099191-cgu.1.rcgu.o" "/var/tmp/portage/media-libs/mesa-9999/work/mesa-9999-abi_x86_32.x86/meson-private/rusttest.sanity.bf099191-cgu.2.rcgu.o" "/var/tmp/portage/media-libs/mesa-9999/work/mesa-9999-abi_x86_32.x86/meson-private/rusttest.sanity.bf099191-cgu.3.rcgu.o" "/var/tmp/portage/media-libs/mesa-9999/work/mesa-9999-abi_x86_32.x86/meson-private/rusttest.sanity.bf099191-cgu.4.rcgu.o" "/var/tmp/portage/media-libs/mesa-9999/work/mesa-9999-abi_x86_32.x86/meson-private/rusttest.2q342kddakz9a0pl.rcgu.o" "-Wl,--as-needed" "-L" "/usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-2d11247ad2092e21.rlib" "/usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-3f4ddc97dc2e2b8b.rlib" "/usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-f82405462986a756.rlib" "/usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libmemchr-7e102d4bb386f25e.rlib" "/usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-6b7de8469acee22c.rlib" "/usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-1f2f4359c54271c9.rlib" "/usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-514d45f94105421d.rlib" "/usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_detect-fd9302e963e0105b.rlib" "/usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-39662e4cbb5f343b.rlib" "/usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libminiz_oxide-d3d423938843589f.rlib" "/usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libadler-db9b40957eb899c6.rlib" "/usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-79795053dc3259ce.rlib" "/usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-6835b48b95346c2b.rlib" "/usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-688e0f765a092900.rlib" "/usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-5e2a591c0dd24aef.rlib" "/usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-d28353d1a6aff89b.rlib" "/usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-b686c46620b68e3c.rlib" "/usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-8cd9ef3e229ed1a5.rlib" "/usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-486966850f786d92.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "/usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/var/tmp/portage/media-libs/mesa-9999/work/mesa-9999-abi_x86_32.x86/meson-private/rusttest" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro,-znow" "-nodefaultlibs" "-m32" "-mfpmath=sse"
  = note: ld.lld: error: /var/tmp/portage/media-libs/mesa-9999/temp/rustc5oecoZ/symbols.o is incompatible with elf32-i386
          ld.lld: error: /var/tmp/portage/media-libs/mesa-9999/work/mesa-9999-abi_x86_32.x86/meson-private/rusttest.sanity.bf099191-cgu.0.rcgu.o is incompatible with elf32-i386
          ld.lld: error: /var/tmp/portage/media-libs/mesa-9999/work/mesa-9999-abi_x86_32.x86/meson-private/rusttest.sanity.bf099191-cgu.1.rcgu.o is incompatible with elf32-i386
          ld.lld: error: /var/tmp/portage/media-libs/mesa-9999/work/mesa-9999-abi_x86_32.x86/meson-private/rusttest.sanity.bf099191-cgu.2.rcgu.o is incompatible with elf32-i386
          ld.lld: error: /var/tmp/portage/media-libs/mesa-9999/work/mesa-9999-abi_x86_32.x86/meson-private/rusttest.sanity.bf099191-cgu.3.rcgu.o is incompatible with elf32-i386
          ld.lld: error: /var/tmp/portage/media-libs/mesa-9999/work/mesa-9999-abi_x86_32.x86/meson-private/rusttest.sanity.bf099191-cgu.4.rcgu.o is incompatible with elf32-i386
          ld.lld: error: /var/tmp/portage/media-libs/mesa-9999/work/mesa-9999-abi_x86_32.x86/meson-private/rusttest.2q342kddakz9a0pl.rcgu.o is incompatible with elf32-i386
          ld.lld: error: /usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-2d11247ad2092e21.rlib(lib.rmeta) is incompatible with elf32-i386
          ld.lld: error: /usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-2d11247ad2092e21.rlib(std-2d11247ad2092e21.std.9601b68c-cgu.0.rcgu.o) is incompatible with elf32-i386
          ld.lld: error: /usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-3f4ddc97dc2e2b8b.rlib(lib.rmeta) is incompatible with elf32-i386
          ld.lld: error: /usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-3f4ddc97dc2e2b8b.rlib(panic_unwind-3f4ddc97dc2e2b8b.panic_unwind.e74d2214-cgu.0.rcgu.o) is incompatible with elf32-i386
          ld.lld: error: /usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-f82405462986a756.rlib(lib.rmeta) is incompatible with elf32-i386
          ld.lld: error: /usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-f82405462986a756.rlib(object-f82405462986a756.object.86afcd81-cgu.0.rcgu.o) is incompatible with elf32-i386
          ld.lld: error: /usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libmemchr-7e102d4bb386f25e.rlib(lib.rmeta) is incompatible with elf32-i386
          ld.lld: error: /usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libmemchr-7e102d4bb386f25e.rlib(memchr-7e102d4bb386f25e.memchr.66a48a28-cgu.0.rcgu.o) is incompatible with elf32-i386
          ld.lld: error: /usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-6b7de8469acee22c.rlib(lib.rmeta) is incompatible with elf32-i386
          ld.lld: error: /usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-6b7de8469acee22c.rlib(addr2line-6b7de8469acee22c.addr2line.8ba115c8-cgu.0.rcgu.o) is incompatible with elf32-i386
          ld.lld: error: /usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-1f2f4359c54271c9.rlib(lib.rmeta) is incompatible with elf32-i386
          ld.lld: error: /usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-1f2f4359c54271c9.rlib(gimli-1f2f4359c54271c9.gimli.4664d47e-cgu.0.rcgu.o) is incompatible with elf32-i386
          ld.lld: error: /usr/lib/rust/1.65.0/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-514d45f94105421d.rlib(lib.rmeta) is incompatible with elf32-i386
          ld.lld: error: too many errors emitted, stopping now (use --error-limit=0 to see all errors)
          clang-15: error: linker command failed with exit code 1 (use -v to see invocation)
          

error: aborting due to previous error

FireBurn commented:
If I just build the 64bit like I used to I then get the header issue:

../mesa-9999/src/compiler/clc/meson.build:33:0: ERROR: File /usr/lib/llvm/15/lib64/clang/15.0.4/include/opencl-c-base.h does not exist.

I'll see about raising a bug for that with the mesa folk, as it's assiming clang is installed on the llvm directory

mattst88 commented:
I'd suggest just not supporting rusticl on 32-bit x86, unless there's a compelling reason.

thesamesam commented:
Thanks, this looks far better now. Obviously it's not going to be merged until it's merged upstream but I think it's in a good state.

darkbasic commented:
@thesamesam it is already merged upstream, just not for AMD (if you have an Intel card you can already use it).
The real blocker IMO is this: https://bugs.gentoo.org/880615

darkbasic commented:
@FireBurn I've removed the duplicate rusticl use flag, please let me know what was your `lpcl` question about.

By the way, rusticl is going to be hitting mesa 22.3 stable next week: https://www.phoronix.com/news/Mesa-22.3-Next-Week

darkbasic commented:
I've removed the `lpcl` use flag (which required the `LP_CL=1` env variable) because apparently the same llvmpipe opencl sw implementation can be accessed via `RUSTICL_ENABLE=lp` whenever you compile mesa with the `rusticl` use flag. You can even use `RUSTICL_DEVICE_TYPE=gpu` to fake it as true-gpu-i-swear-its-non-sw implementation to work around application blacklistings.
I'd say this is good enough to drop the `lpcl` use flag.

darkbasic commented:
We should be ready to merge: please check again and let me know if there are any further issues.

AuroransSolis commented:
There may be an issue with building with the both the `abi_x86_32` and `rusticl` USE flags. I'm not sure if it's something wrong on my end or if it's a conflict between assumed default arches between the GCC linker (i386 default, I think?) and Rust (i686 default, I think?) and maybe needing some extra work to match those.

darkbasic commented:
What's holding this back? All blockers have been solved...

stefson commented:
Any news on this pullrequest? Might be good sport to try to compile it for mesa-9999 at least, wouldn't it?

mattst88 commented:
> Any news on this pullrequest? Might be good sport to try to compile it for mesa-9999 at least, wouldn't it?

I simply don't have time or bandwidth to take on extra maintenance burdens on top of my existing responsibilities in Gentoo.

darkbasic commented:
> I simply don't have time or bandwidth to take on extra maintenance burdens on top of my existing responsibilities in Gentoo.

Please let me know if I can help you: my hardware is limited to ppc64 + radeonsi but that I will keep testing. I think that adding it only to the live ebuild might be a good compromise for the moment (especially since ATM it only works on Intel GPUs with mainline code): low maintenance burden plus that what's most of rusticl users will use anyway.

darkbasic commented:
Upstream bumped meson requirements for rusticl to 1.0.0, but since no one is rushing to get this merged I won't bump the dependency (it will probably get stabilized soon anyway). Also there are some new issues with the latest code and runtime dependencies detection, I'm working with the upstream to get them fixed.

cyro666 commented:
I tried your branch for mesa-9999, but abi_x86_32 is holding me back as well, it errors out when trying to compile the 32 bit part. I don't even remember why I have this defined for mesa and a lot of other stuff, but it's probably wine (and wine is working on getting rid of 32-bit host stuff anyways, it'll just take some time).

In any case, just wanted to chime in, because I would very much like to an option for building rusticl on Gentoo as well.

darkbasic commented:
Unfortunately 32 bit is known to not work at the moment and should be masked if we decide to get this merged. Also the radeonsi branch needs the following patch at the moment:

```
diff --git a/src/compiler/clc/clc_helpers.cpp b/src/compiler/clc/clc_helpers.cpp
index 80bc84831e4..34d156bf227 100644
--- a/src/compiler/clc/clc_helpers.cpp
+++ b/src/compiler/clc/clc_helpers.cpp
@@ -39,6 +39,7 @@
 #include <llvm-c/Target.h>
 #include <LLVMSPIRVLib/LLVMSPIRVLib.h>
 
+#include <clang/Config/config.h>
 #include <clang/Driver/Driver.h>
 #include <clang/CodeGen/CodeGenAction.h>
 #include <clang/Lex/PreprocessorOptions.h>
@@ -866,7 +867,7 @@ clc_compile_to_llvm_module(LLVMContext &llvm_ctx,
    // because we might have linked clang statically.
    auto libclang_path = fs::path(LLVM_LIB_DIR) / "libclang.so";
    auto clang_res_path =
-      fs::path(clang::driver::Driver::GetResourcesPath(libclang_path.string())) / "include";
+      fs::path(clang::driver::Driver::GetResourcesPath(libclang_path.string(), CLANG_RESOURCE_DIR)) / "include";
 
    c->getHeaderSearchOpts().UseBuiltinIncludes = true;
    c->getHeaderSearchOpts().UseStandardSystemIncludes = true;

```

stefson commented:
is there an upstream bug/patch for the clc_helpers.cpp file?

darkbasic commented:
There is no need for one: I've discussed with Karol Herbst about the issue and he is not super happy about the fix so he is still trying to find a better solution. He should probably setup Gentoo prefixes because it's not the first time it regresses.

darkbasic commented:
I removed my bindgen package since it has already been added to the Gentoo repos by someone else in the meantime.

darkbasic commented:
radeonsi support has been merged in upstream mesa, so at this point I would **REALLY** love to see this merged (I don't think I'm the only one). I've cleaned and rebased everything and a lot has changed since this PR was originally opened, in fact the changeset is pretty small now (could be even smaller if you agree to get rid of clover, which is going to be stripped out of mesa soon anyway). Also both me and and @FireBurn already maintain our own rusticl mesa ebuilds and we will still be around if anything breaks.

NekoEd commented:
Adding my voice to this since it's been a few weeks since the last activity; it would be great to get this merged given Clover's (OpenCL 1.2) deprecation and rusticl's considerable increase (OpenCL 3.0) in functionality with the new AMDGPU support. Hardly anything supports OpenCL 1.2 anymore, and ROCm hardly supports any cards, so this is a few steps beyond "nice to have" by this point.

cyro666 commented:
Besides, as mentioned - this is upstream now. I understand people are probably busy, but this really should get merged eventually.

darkbasic commented:
Also ROCm hardly works on anything but x86, even if you have a supported GPU.

P.S.
I've just rebased the PR.

mattst88 commented:
I'm planning to merge this for Mesa 23.1.

mattst88 commented on media-libs/mesa/mesa-9999.ebuild:
```diff
@@ -316,6 +316,12 @@ pkg_setup() {
 
+src_prepare() {
+	# Temporary workaround: https://gitlab.freedesktop.org/mesa/mesa/-/issues/7717#note_1832122
+	use rusticl && eapply "${FILESDIR}/clang_resource_dir.patch"
```
Is there any danger in applying this unconditionally?

darkbasic replied:
I don't think so.

mattst88 commented on media-libs/mesa/mesa-9999.ebuild:
```diff
@@ -32,7 +32,7 @@ done
 IUSE="${IUSE_VIDEO_CARDS}
 	cpu_flags_x86_sse2 d3d9 debug gles1 +gles2 +llvm
-	lm-sensors opencl osmesa +proprietary-codecs selinux
+	lm-sensors opencl clover rusticl osmesa +proprietary-codecs selinux
```
This list should be alphabetized.

mattst88 commented on media-libs/mesa/metadata.xml:
```diff
@@ -11,7 +11,9 @@
-    <flag name="opencl">Enable the Clover Gallium OpenCL state tracker.</flag>
+    <flag name="opencl">Enable the Rusticl or Clover Gallium OpenCL state tracker.</flag>
+    <flag name="rusticl">Enable the Rusticl OpenCL state tracker.</flag>
+    <flag name="clover">Enable the Clover Gallium OpenCL state tracker.</flag>
```
Can we drop `clover` entirely? I'm okay with that. I don't think it can be used for really anything. That would let us get rid of both `IUSE=rusticl` as well, and just have rusticl controlled by `IUSE=opencl`.

darkbasic replied:
I'm in favor of dropping clover, it will make the ebuild simpler. The only downside is that R600 users will loose OpenCL for a while (rusticl is planned for R600 but still not working). I guess they can use mesa 23.0 until that happens.

mattst88 replied:
Yeah, I think that's fine. I expect the number of R600 Gentoo users using OpenCL 1.1 is approaching zero :)

mattst88 commented on media-libs/mesa/mesa-9999.ebuild:
```diff
@@ -184,6 +184,10 @@ BDEPEND="
 	${PYTHON_DEPS}
 	opencl? (
 		>=sys-devel/gcc-4.6
```
I'm pretty sure this requirement was something to do with some C++ madness in Clover. I'm going to speculatively remove it.

mattst88 commented on media-libs/mesa/mesa-9999.ebuild:
```diff
@@ -183,7 +183,10 @@ DEPEND="${RDEPEND}
+		>=virtual/rust-1.62.0
+		>=dev-util/bindgen-0.58.0
+		dev-util/spirv-tools
+		dev-util/spirv-llvm-translator
```
The [Rusticl documentation](https://docs.mesa3d.org/rusticl.html) says the build dependencies are

>- rustc
>- rustfmt (highly recommended, but only required for CI builds or when authoring patches)
>- bindgen
>- [LLVM](https://github.com/llvm/llvm-project/) built with libclc and -DLLVM_ENABLE_DUMP=ON.
>- [SPIRV-Tools](https://github.com/KhronosGroup/SPIRV-Tools)
>- [SPIRV-LLVM-Translator](https://github.com/KhronosGroup/SPIRV-LLVM-Translator) for a libLLVMSPIRVLib.so matching your version of LLVM, i.e. if you’re using LLVM 15 (libLLVM.so.15), then you need a libLLVMSPIRVLib.so.15.

The dependency on `dev-util/spirv-llvm-translator` is certainly wrong:

````
/var/tmp/portage/media-libs/mesa-9999/image # scanelf -n ./usr/lib64/libRusticlOpenCL.so
 TYPE   NEEDED FILE 
ET_DYN libLLVM-15.so,libz.so.1,libzstd.so.1,libdrm.so.2,libexpat.so.1,libclang-cpp.so.g1,libLLVMSPIRVLib.so.15,libSPIRV-Tools-opt.so,libSPIRV-Tools.so,libSPIRV-Tools-link.so,libstdc++.so.6,libm.so.6,libgcc_s.so.1,libc.so.6,ld-linux-x86-64.so.2 ./usr/lib64/libRusticlOpenCL.so 
````

It should be in `DEPEND` and `RDEPEND`. I believe it should **not** be in `BDEPEND`, since I don't see any invocations of `llvm-spirv` in the `build.log`:

````
/var/tmp/portage/media-libs/mesa-9999 # grep llvm-spirv temp/build.log
<empty>

````

Additionally, it needs to be linked against the same version of LLVM otherwise used by Mesa.

>The minimum versions to build Rusticl are:
>
> - Rust: 1.59

I don't know why you chose `>=virtual/rust-1.62.0`, but that's fine with me. 1.66 is stable.

> - Meson: 1.0.0

`>=dev-util/meson-1.0.0` definitely needs to be specified as a dependency in `BDEPEND`. `meson.eclass` only requires 0.62.2.

> - Bindgen: 0.58.0

I tested with bindgen-0.56.0 and it failed. bindgen-0.58 isn't in ::gentoo, so I assume this is fine.

> - LLVM: 11.0.0 (recommended: 15.0.0)

We only allow `sys-devel/llvm:15` or `sys-devel/llvm:16` in `mesa-*.ebuild`, so this is good.

> - SPIRV-Tools: any version (recommended: v2022.3)

We package the Vulkan-SDK tags and not the date-based tags, so I don't know how this compares exactly. `dev-util/spirv-tools-1.3.243`'s `/usr/lib64/pkgconfig/SPIRV-Tools.pc` says `Version: 2023.2.1` so it's definitely new enough. I suggest simply requiring `>=dev-util/spirv-tools-1.3.243` and not worrying about the precise lower-bound.

But, like `dev-util/spirv-llvm-translator` this should be in `DEPEND` and `RDEPEND` (and probably **not** in `BDEPEND` because I see no invocations of `spirv-*` programs in the `build.log`). Note that in the `scanelf` output above that Rusticl is linking with `libSPIRV-Tools-opt.so,libSPIRV-Tools.so,libSPIRV-Tools-link.so`.







 I will handle these updates. Thank you for your contribution, and I hope that you'll contribute in the future.

At the same time, I would like to take this opportunity say something to those that thought that all that was needed from my side was to press a button:

- It was not.
- It almost never is.
- Please be patient.
- Your pings do not help.
- They just add stress to a volunteer who is already doing much more than he wants to or feels is sustainable.
- If you want to see something happen sooner, please try to take something else off my plate.

darkbasic replied:
> I don't know why you chose >=virtual/rust-1.62.0

I honestly don't remember. Probably because that was what I initially tested it against and since it was already stable it didn't matter that much.

> >=dev-util/meson-1.0.0 definitely needs to be specified as a dependency

It's one of those deps that got upgraded recently because initially it required a much older version.

> I will handle these updates

Thanks!

> Thank you for your contribution, and I hope that you'll contribute in the future.

Thank you for your review, it's much appreciated!

> At the same time, I would like to take this opportunity say something to those that thought that all that was needed from my side was to press a button

I don't think that anybody thought that a press of button was all that was needed and the long wave of iterations this PR has gone through is the proof that things are never that simple. I personally just wanted to reassure you that I'm actually using this thing and I won't disappear once the PR gets merged.

> They just add stress to a volunteer who is already doing much more than he wants to or feels is sustainable

I know the feeling, I'm currently struggling to succesfully compile latest Chromium 113 on ppc while I'm bisecting the kernel because 6.3 decided to read random memory pages and fails to boot, latest vscode stopped working against my custom ppc64 electron build and qtwebengine regressed as well. All while trying to get some actual work done, which heavily relies on having working Chromium/electron ebuilds since I'm a full stack Javascript developer.

mattst88 commented on media-libs/mesa/mesa-9999.ebuild:
```diff
@@ -386,10 +394,14 @@ multilib_src_configure() {
-		-Dgallium-opencl="$(usex opencl icd disabled)"
-	)
+	if use opencl; then
+		PKG_CONFIG_LIBDIR="$(get_llvm_prefix)/$(get_libdir)/pkgconfig:${ESYSROOT}/usr/$(get_libdir)/pkgconfig"
```
After quite a bit of debugging, I think this is wrong. `PKG_CONFIG_LIBDIR` is apparently the primary path, and `PKG_CONFIG_PATH` is the secondary path.

If you set `PKG_CONFIG_PATH` you don't have to include `${ESYSROOT}/usr/$(get_libdir)/pkgconfig` because you're not overwriting the primary path (which already contains that).

mattst88 commented:
I'm wiring up `intel_clc` support for v23.1.0 as well (it is necessary for Vulkan ray tracing support on Intel), and since it also depends on `spirv-llvm-translator` and there are some difficulties around that dependency, I'm going to wait to push both `intel_clc` and `rusticl` support at the same time.

I'm currently waiting on https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/22968 to land upstream so I don't have to then make another commit in `::gentoo` to remove the patch from `mesa-9999.ebuild`.
